modules/social_features/social_core/src/Plugin/Validation/Constraint/ExternalIdentifierUniqueExternalIdConstraintValidator.php

```diff
@@ -0,0 +1,186 @@
+ *   | Group:1 | field_external_id_1 | 123         | consumer:1     |
+ *
+ * Example 6: This situation is allowed as External ID is unique per external
+ *            owner.
+ *
+ *   | Entity | Field               | External ID | External owner |
+ *   | ------ | ------------------- | ----------- | -------------- |
+ *   | User:1 | field_external_id_1 | 123         | consumer:1     |
+ *   | User:1 | field_external_id_1 | 123         | consumer:2     |
+ *   | User:1 | field_external_id_1 | 123         | owner:1        |
+ */
+class ExternalIdentifierUniqueExternalIdConstraintValidator extends ConstraintValidator implements ContainerInjectionInterface {
+
+  const EXTERNAL_IDENTIFIER_FIELD_TYPE = 'social_external_identifier';
+
+  /**
+   * Constructs a new constraint validator.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager service.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entityFieldManager
+   *   The entity field manager.
+   */
+  public function __construct(
+    protected EntityTypeManagerInterface $entityTypeManager,
+    protected EntityFieldManagerInterface $entityFieldManager
+  ) {
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container): static {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('entity_field.manager')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate(mixed $item, Constraint $constraint) {
+    if (!$constraint instanceof ExternalIdentifierUniqueExternalIdConstraint) {
+      return;
+    }
+
+    // Empty constraint is handled by
+    // ExternalIdentifierEmptySubfieldsConstraint.
+    if ($item->isEmpty()) {
+      return;
+    }
+
+    $entity = $item->getEntity();
+    $entity_type = $entity->getEntityTypeId();
+
+    // Check per each field of type "social_external_identifier" in given
+    // entity type, if external_id is unique per external owner.
+    foreach ($this->getFieldNamesByFieldType($entity_type) as $field) {
+      $query = $this->entityTypeManager->getStorage($entity_type)->getQuery();
+      $query->condition($field . '.external_id', $item->external_id);
+      $query->condition($field . '.external_owner_target_type', $item->external_owner_target_type);
+      $query->condition($field . '.external_owner_id', $item->external_owner_id);
+      // Exclude existing current entity.
+      if (!$entity->isNew()) {
+        $query->condition('id', $entity->id(), '<>');
+      }
+      $query->accessCheck(FALSE);
+      $result = $query->execute();
+
+      if (!empty($result)) {
+        $this->context->addViolation($constraint->externalIdNotUniqueMessage, [
+          '%external_id' => $item->external_id,
+          // @todo Do we even want to expose internal owner target type and id?
+          '%external_owner_target_type' => $item->external_owner_target_type,
```

Comment:
Is it ok to expose which external owner entity has already used external identifier id value in error message?

Reply:
As discussed with @bramtenhove we will not expose internal owner and target type. Based on this some changes has been applied.

Message
`External identifier id should be unique. External identifier id "%external_id" is already used, with external owner "%external_owner_target_type" of id "%external_owner_id".` is changed to `The external identifier ID must be unique. The External ID "%external_id" is already in use.`

and violation is changed from

```
        $this->context->addViolation($constraint->externalIdNotUniqueMessage, [
          '%external_id' => $item->external_id,
          '%external_owner_target_type' => $item->external_owner_target_type,
          '%external_owner_id' => $item->external_owner_id,
        ]);
```
to
```
        $this->context->addViolation($constraint->externalIdNotUniqueMessage, [
          '%external_id' => $item->external_id,
        ]);
```